procedures/statement_performance_analyzer.sql

```diff
@@ -0,0 +1,792 @@
+             ...
+            '
+    SQL SECURITY INVOKER
+    NOT DETERMINISTIC
+    CONTAINS SQL
+BEGIN
+    DECLARE v_table_exists, v_tmp_digests_table_exists, v_custom_view_exists ENUM('', 'BASE TABLE', 'VIEW', 'TEMPORARY') DEFAULT '';
+    DECLARE v_this_thread_enabled ENUM('YES', 'NO');
+    DECLARE v_force_new_snapshot BOOLEAN DEFAULT FALSE;
+    DECLARE v_digests_table VARCHAR(133);
+    DECLARE v_quoted_table, v_quoted_custom_view VARCHAR(133) DEFAULT '';
+    DECLARE v_table_db, v_table_name, v_custom_db, v_custom_name VARCHAR(64);
+    DECLARE v_digest_table_template, v_checksum_ref, v_checksum_table text;
+    -- Maximum supported length for MESSAGE_TEXT with the SIGNAL command is 128 chars.
+    DECLARE v_error_msg VARCHAR(128);
+
+
+    -- Don't instrument this thread
+    SELECT INSTRUMENTED INTO v_this_thread_enabled FROM performance_schema.threads WHERE PROCESSLIST_ID = CONNECTION_ID();
+    CALL sys.ps_setup_disable_thread(CONNECTION_ID());
+
+    -- Temporary table are used - disable sql_log_bin if necessary to prevent them replicating
+    SET @log_bin := @@sql_log_bin;
+    IF (@log_bin = 1) THEN
+        SET sql_log_bin = 0;
+    END IF;
+
+
+    -- Set configuration options
+    IF (@sys.statement_analyzer_limit IS NULL) THEN
+        SET @sys.statement_analyzer_limit = sys.sys_get_config('statement_analyzer_limit', '100');
+    END IF;
+    IF (@sys.debug IS NULL) THEN
+        SET @sys.debug                    = sys.sys_get_config('debug'                   , 'OFF');
+    END IF;
+
+
+    -- If in_table is set, break in_table into a db and table component and check whether it exists
+    -- in_table = NOW() is considered like it's not set.
+    IF (in_table = 'NOW()') THEN
+        SET v_force_new_snapshot = TRUE,
+            in_table             = NULL;
+    ELSEIF (in_table IS NOT NULL) THEN
+        IF (NOT INSTR(in_table, '.')) THEN
+            -- No . in the table name - use current database
+            -- DATABASE() will be the database of the procedure
+            SET v_table_db   = DATABASE(),
+                v_table_name = in_table;
+        ELSE
+            SET v_table_db   = SUBSTRING_INDEX(in_table, '.', 1);
+            SET v_table_name = SUBSTRING(in_table, CHAR_LENGTH(v_table_db)+2);
+        END IF;
+
+        SET v_quoted_table = CONCAT('`', v_table_db, '`.`', v_table_name, '`');
+
+        IF (@sys.debug = 'ON') THEN
+            SELECT CONCAT('in_table is: db = ''', v_table_db, ''', table = ''', v_table_name, '''') AS 'Debug';
+        END IF;
+
+        IF (v_table_db = DATABASE() AND (v_table_name = 'tmp_digests' OR v_table_name = 'tmp_digests_delta')) THEN
+            SET v_error_msg = CONCAT('Invalid value for in_table: ', v_quoted_table, ' is reserved table name.');
+            SIGNAL SQLSTATE '45000'
+               SET MESSAGE_TEXT = v_error_msg;
+        END IF;
+
+        CALL table_exists(v_table_db, v_table_name, v_table_exists);
```

Comment:
Prefix this with "sys.table_exists", these can be called from outside of the sys schema.